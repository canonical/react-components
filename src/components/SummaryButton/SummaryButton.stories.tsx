import React from "react";
import { Meta, StoryObj } from "@storybook/react";

import SummaryButton from "./SummaryButton";

const doNothing = () => {};

const meta: Meta<typeof SummaryButton> = {
  component: SummaryButton,
  tags: ["autodocs"],
};

export default meta;

type Story = StoryObj<typeof SummaryButton>;

export const Default: Story = {
  render: () => (
    <SummaryButton
      summary="Showing 15 out of 100 items."
      label="Show more"
      onClick={doNothing}
    />
  ),

  name: "Default",
};

/**
 * Summary text is optional.
 */
export const ActionOnly: Story = {
  render: () => <SummaryButton label="Show more items" onClick={doNothing} />,
  name: "Action only",
};

/**
 * Action is optional as well. When `onClick` prop is not provided component will render just the summary text.
 */
export const SummaryOnly: Story = {
  render: () => (
    <SummaryButton
      summary="Showing 100 out of 100 items."
      label="Show more items"
      onClick={doNothing}
    />
  ),
  name: "Summary only",
};

/**
 * When data is loaded asynchronusly `isLoading` prop should be set to `true` to shows a spinner animation in place of the action button.
 */
export const Loading: Story = {
  render: () => (
    <SummaryButton
      summary="Showing 15 out of 100 items."
      label="Show more"
      isLoading
      onClick={doNothing}
    />
  ),

  name: "Loading",
};
