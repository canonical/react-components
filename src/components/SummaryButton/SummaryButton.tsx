import React from "react";
import type { MouseEventHandler } from "react";
import classNames from "classnames";

import ActionButton from "../ActionButton";

import type { ClassName } from "types";

export type Props = {
  /**
   * Optional class(es) to pass to the wrapping element.
   */
  className?: ClassName;
  /**
   * Whether the summary button is loading.
   */
  isLoading?: boolean;
  /**
   * The label of the summary button.
   */
  label: string;
  /**
   * Function to handle clicking the summary button.
   */
  onClick: MouseEventHandler<HTMLButtonElement>;
  /**
   * The summary content.
   */
  summary?: string;
};

const SummaryButton = ({
  className,
  isLoading,
  summary,
  label,
  onClick,
}: Props): JSX.Element => (
  <small className={className}>
    {summary && <span className="u-text--muted">{summary}</span>}
    {onClick && (
      <ActionButton
        className={classNames("is-small", "is-dense", {
          "is-inline": summary,
        })}
        onClick={onClick}
        loading={isLoading}
        disabled={isLoading}
      >
        {label}
      </ActionButton>
    )}
  </small>
);

export default SummaryButton;
