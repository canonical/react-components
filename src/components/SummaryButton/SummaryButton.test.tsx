import React from "react";
import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";

import SummaryButton from "./SummaryButton";

describe("<SummaryButton />", () => {
  // snapshot tests
  it("renders and matches the snapshot", () => {
    const { container } = render(
      <SummaryButton
        summary="Showing some items"
        label="Show more"
        onClick={jest.fn()}
      />,
    );

    expect(container.firstChild).toMatchSnapshot();
  });

  // unit tests
  it("renders a spinner icon", () => {
    const { container } = render(
      <SummaryButton label="Show more" onClick={jest.fn()} isLoading />,
    );

    expect(container.querySelector("i.u-animation--spin")).toBeInTheDocument();
  });

  it("can handle click events", async () => {
    const onClick = jest.fn();
    render(<SummaryButton label="Show more" onClick={onClick} />);

    await userEvent.click(screen.getByRole("button", { name: "Show more" }));

    expect(onClick).toHaveBeenCalled();
  });
});
