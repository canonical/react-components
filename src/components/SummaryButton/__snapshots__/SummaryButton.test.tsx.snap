// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SummaryButton /> renders and matches the snapshot 1`] = `
<small>
  <span
    className="u-text--muted"
  >
    Showing some items
  </span>
  <ActionButton
    className="is-small is-dense is-inline"
    onClick={[Function]}
  >
    Show more
  </ActionButton>
</small>
`;
